Add rake task to report the number of draft...

deborahchua opened the pull request:
...publications by organisation within a given date range.

This rake task will generate a report to CSV as well as output in
the console.
Running this task weekly will mean we can monitor and see if there
is an increase in draft publications.

Trello card: https://trello.com/c/rLXkQ9Nj/835-find-out-how-many-drafts-exist-in-the-system-at-a-given-time

thomasleese commented on lib/tasks/reporting.rake:
```diff
@@ -80,6 +80,47 @@ namespace :reporting do
+
+    CSV_HEADERS = ["Lead publishing organisation", "Number of draft publications"].freeze
+
+    CSV.open(path, 'wb') do |csv|
```
You can actually do `CSV.open(path, "wb", headers: CSV_HEADERS, write_headers: true)` here to avoid the line below.

thomasleese commented on lib/tasks/reporting.rake:
```diff
@@ -80,6 +80,47 @@ namespace :reporting do
+    if args[:start_date].present? && args[:end_date].present?
+      start_date = Time.parse(args[:start_date])
+      end_date = Time.parse(args[:end_date])
+    end
```
Should we raise an error here if these arguments are missing? Perhaps like:

```ruby
raise "Missing start_time or end_time" unless args[:start_date].present? && args[:end_date].present?
```

and then we don't need the `if` block.

thomasleese commented on lib/tasks/reporting.rake:
```diff
@@ -80,6 +80,47 @@ namespace :reporting do
+
+      organisations = Organisation.order(slug: :asc)
+
+      organisations.each do |organisation|
```
Although it might not make a difference here, `.find_each` can often be quicker as it avoids allocating a new object for each row in the database and reuses the same one each time.

deborahchua commented:
Thanks @thomasleese ! I have implemented your suggestions and moved the code into its own class in https://github.com/alphagov/whitehall/pull/4667/commits/601976abae5f51a11d22311c967609961a92ca8a
